cyfronet-fid/marketplace
# [#1966] Expose user_secrets in ordering API

The user_secrets hash values are obfuscated, excluding the ones which
were updated by the call (i.e. in method update).

Setting `user_secrets: {}` in policy's permitted_attributes allows any
key.

Ensure user_secrets are obfuscated in events.

# How to test

API-key: `ck_B7F_WxbZQ9MghtdPm`
oms_id: `1`

1. Log in as a user under https://pr-1992.docker-fid.grid.cyf-kr.edu.pl/.
2. Create a project with a voucherable project_item (for service EC3).
3. Navigate to the Ordering API Swagger-UI: https://pr-1992.docker-fid.grid.cyf-kr.edu.pl/api_docs/swagger/index.html?urls.primaryName=Ordering%20API%20V1%20Docs and authorize with the API-key above.
4. Determine the created project_item_id and project_id (for example using /events? endpoint). All the urls are assumed to be prefixed with `/oms/{oms_id}`.
5. `GET /projects/{project_id}/project_item/{project_item_id}`. See if it renders an empty user_secrets, i.e. `user_secrets: {}`.
6. `PATCH /projects/{project_id}/project_item/{project_item_id}` with payload `{"user_secrets":{"one": "value", "two": 2}}`. You should get 400.
7. `PATCH /projects/{project_id}/project_item/{project_item_id}` with payload `{"user_secrets":{"one": "value"}}`. You should get 200 including the value of the user_secret.
5. `GET /projects/{project_id}/project_item/{project_item_id}`. See if it renders obfuscated user_secrets, i.e. `user_secrets: {one: "<OBFUSCATED>"}`.
7. `PATCH /projects/{project_id}/project_item/{project_item_id}` with payload `{"user_secrets":{"voucher_id": "value"}}`. You should get 200 including the value of the `user_secret = {"one": "<OBFUSCATED>", "voucher_id": "value"}`.
5. `GET /projects/{project_id}/project_item/{project_item_id}`. See if it renders obfuscated user_secrets, i.e. `user_secrets: {one: "<OBFUSCATED>", voucher_id: "<OBFUSCATED>"}`.
6. `GET /events?from_timestamp=2020-01-01T00:00:00Z`, ensure that the returned events don't include user_secrets values but `<OBFUSCATED>`.
7. Go as a user to the project_item and ensure that she sees Voucher ID=value.
